Artifact generation KEP (#858)
kubernetes/enhancements: keps/sig-cluster-lifecycle/20190227-artifact-generation.md

---

# package-generation

## Table of Contents

- [Title](#title)
  - [Table of Contents](#table-of-contents)
  - [Release Signoff Checklist](#release-signoff-checklist)
  - [Summary](#summary)
  - [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals](#non-goals)
  - [Proposal](#proposal)
    - [User Stories [optional]](#user-stories-optional)
        - [Story 1](#story-1)
        - [Story 2](#story-2)
        - [Story 3](#story-3)
    - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
    - [Risks and Mitigations](#risks-and-mitigations)
  - [Design Details](#design-details)
    - [Test Plan](#test-plan)
    - [Graduation Criteria](#graduation-criteria)
    - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
    - [Version Skew Strategy](#version-skew-strategy)
  - [Implementation History](#implementation-history)
  - [Drawbacks [optional]](#drawbacks)
  - [Alternatives [optional]](#alternatives)
  - [Infrastructure Needed [optional]](#infrastructure-needed)

[Tools for generating]: https://github.com/ekalinin/github-markdown-toc

## Release Signoff Checklist

> **ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)

Author: removing.

[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
[kubernetes/website]: https://github.com/kubernetes/website

## Summary

Implementation of a standard, community supported and maintained process of generating and building rpm and deb packages.

## Motivation

Today, the tooling used to build kubernetes packages is separate from kubernetes itself. This can cause downstream divergence, and cannot be easily maintained by the kubernetes community as a whole. 

The primary goal of this proposal is to simplify the lifecycle management of packaging by co-locating the sources and tooling for generating packages into the kubernetes/kubernetes repository. 

### Goals

To create a maintainable and configurable method of building kubernetes artifacts supported and maintained by the community.

- Move packaging tools out of kubernetes/release and into kubernetes/kubernetes repository
- Enable customization for downstream packaging 
    - Including the ability to bring their own source code/binaries and changelogs 
    - Stop hard-coding the binary download URL to dl.k8s.io

> - Enable community governance over packaging of the project.

Author: sig-release. I can state that more clearly.

> - Enable community governance over packaging of the project.

Reviewer: Thanks, that would benefit people looking for guidance in this document.

- Tightly couple the versioning of packing and the kubernetes/kubernetes release.

> - Ensure that the same packages that are run in CI are the ones that are released.

Reviewer 2: not sure if this should be part of this KEP or the other KEP about publishing.

i fixed a bug yesterday related to CI packages:
https://github.com/kubernetes/test-infra/pull/11534

the CI packages are generated by bazel and pushed to GCS by a post-submit job.
to preserve the existing CI behavior and if bazel is no longed used for building debs/rpms, new prow jobs have to be created that exercise the new package generation tooling.

Author: I don't know that this would be true. Bazel could/should in fact call this to package the built binary. Meaning Bazel could continue to push to GCS afterwards. Also, I think that in the future per the other KEP mentioned we will start to push to CNCF so that part will be changing anyhow if I understand your question correctly.

- Dedupe the generation of packages (THERE CAN BE ONLY ONE!) 
- Have one config file that is the source of truth for packaging details for both .rpm and .deb packages.
- Artifact generation is independent of source code compilation. 

### Non-Goals

- A change in the release process. see this [kep](https://github.com/kubernetes/enhancements/pull/843)
    - For example, package signing and distribution are not covered in this document and are covered in the above KEP.

- To fundamentally change the packages and their dependencies. All packages should remain consistent with what we build today. 

## Proposal

As outlined in the following issue:

> Base ticket: https://github.com/kubernetes/kubernetes/issues/71677

Author: not sure I understand what you asking.

Move all packaging into kubernetes/kubernetes repo and create the following 3 building blocks to support it.
Create a configuration yaml file to enable the configuring of the built packages.
Create a Go lang app to parse yaml config and submit simple commands to the container.
Build inside a docker container based off FPM.

### User Stories

#### Story 1

As a developer I would like to be able to install the entire set of alpha or beta (.deb/.rpm) packages to stand up a development cluster.  These packages should be equivalent to the future release artifacts.  

#### Story 2

> As a third party vendor, I will be able to package kubernetes using the same tooling that kubernetes’ uses. 

Reviewer 2: currently this tool is bazel.
bazel can still be used by third parties, except that we want a separate tool to decouple the process.
existing binaries -> DEBs/ RPMs.

Reviewer: ```suggestion
As a third party vendor, I would like to be able to package kubernetes deliverable artifacts using the same tooling that kubernetes’ uses.
```

@neolit123 point of my suggestion is to have tooling that can generate customized deliverable artifacts without modifying or wrapping them.

#### Story 3

As a release manager I would like to enable simple and easy package generation that the community can manage. 

### Implementation Details/Notes/Constraints [optional]

- Use an fpm based container to build packages
    - Nfpm was originally investigated but lacked features
- Use yaml to configure

> - Create a go binary that can consume the yaml, build the right file structure and issue commands to the image to build the image.

Reviewer: this might require a separate KEP

Author: I don't think so really, the go binary is the glue between the yaml config file and the docker container that builds it. without it you have two things that don't know about each other and we would be stuck shimming this into bazel maybe?

Reviewer 3: I agree - we should not do a separate KEP for tooling we need to build packages.

Reviewer: If the tool is single purpose, I agree.

Author: Let me know what you think. I tried to make this more clear @frapposelli

### Risks and Mitigations

Risk: Confusion as to what to use when. Mitigation: Coordination across groups to ensure that the changes made in k/k are consumed by the release team. 
## Design Details

### Test Plan

Ideally these new packages will get tested inside the test pipeline. When KIND goes to run kubeadm/kubelet/kubectl/etc tests, it should install the packages generated by these tools as they will be contained within the kubernetes/kubernetes repository.

> This will give us clear paths to the package generated and the package consumed when testing enabling a more confident release.

Author: I will update this to be more clear.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md

### Graduation Criteria

N/A

#### Examples

1:
An independent developer wanting to generate packages in their own fork would be able to edit the config yaml in their fork, override a dependency name and then issue a command to the go packaging binary to generate the packages.
`[go packing binary] --config=myconfig.yaml`

2:
Bazel will be able to update the configuration yaml to dynamically update the configuration before simply executing commands directly against the go packing binary.

> 

Reviewer 2: should this section be N/A and the contents be move as a User Story?

Author: I'm not really sure. I felt like the examples where a bit of a rehash of user stories as well as implementation details was a rehash of proposal. Sorry, this is my First KEP, I'm up for whatever you all think is best. I can do that. These are worded slightly differently...

---

> title: Package Generation

Author: I will update this as soon as I can. will update to `Artifact Generation`

authors:
  - "@klaven"
  - "@ncdc"
owning-sig: sig-cluster-lifecycle
participating-sigs:
  - sig-cluster-lifecycle
  - sig-release
reviewers:
  - "@ncdc"
  - "@hhorl"
  - “@neolit123”
  - “@ixdy”
approvers:
  - "@timothysc"
editor: "@Klaven"
creation-date: 2019-02-15
last-updated: yyyy-mm-dd
status: implementable
see-also:
  - "https://github.com/kubernetes/enhancements/pull/843"
---

# package-generation

## Table of Contents

> - [Title](#title)

Reviewer 4: I think the TOC is out of sync.

Author: will check it when I get a chance and update as needed.

- To fundamentally change the packages and their dependencies. All packages should remain consistent with what we build today. 

## Proposal

As outlined in the following issue:
Base ticket: https://github.com/kubernetes/kubernetes/issues/71677
Move all packaging into kubernetes/kubernetes repo and create the following 3 building blocks to support it.
Create a configuration yaml file to enable the configuring of the built packages.
Create a Go lang app to parse yaml config and submit simple commands to the container.
Build inside a docker container based off FPM.

### User Stories

#### Story 1

As a developer I would like to be able to install the entire set of alpha or beta (.deb/.rpm) packages to stand up a development cluster.  These packages should be equivalent to the future release artifacts.  

#### Story 2

As a third party vendor, I will be able to package kubernetes using the same tooling that kubernetes’ uses. 

#### Story 3

As a release manager I would like to enable simple and easy package generation that the community can manage. 

### Implementation Details/Notes/Constraints [optional]

- Use an fpm based container to build packages

>     -Nfpm was originally investigated but lacked features

Reviewer 4: ```suggestion
    - Nfpm was originally investigated but lacked features
```

- Use yaml to configure
- Create a go binary that can consume the yaml, build the right file structure and issue commands to the image to build the image.

### Risks and Mitigations

> Rist: Confusion as to what to use when. Mitigation: Coordination across groups to ensure that the changes made in k/k are consumed by the release team. 

Reviewer 4: ```suggestion
Risk: Confusion as to what to use when. Mitigation: Coordination across groups to ensure that the changes made in k/k are consumed by the release team. 
```
